### PR TITLE
Grant Application: Roloi - XCM Payment Automation

### PR DESCRIPTION
**Pull request by pcorrado-np**
#### Project Abstract

Our project enables the automation of ink! smart contract transactions in a recurring way. Using XCM, we can schedule transactions on Shiden (Astar) with the assistance of Turing's (OAK) functionalities.  

We will deliver a generic project as a public good that teams can use to automate cross-chain transactions:
- A **Typescript Next.js UI** with abstractions that simplify the most tricky parts for developers such as Accounts connection, Proxy Accounts, and XCM messages creation.
- An **ink! Smart Contract** that handles token transfers between Shiden accounts. 

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [x] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`roloi-xcm-payment-automation.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. Here is some quick feedback:
- the image linked in your application currently doesn't work for me.
- Usually, for UI-focused milestones, we ask teams to already provide some mock-ups or initial designs
- I would recommend mostly focusing on the software that you need to develop as part of the project and not so much any specific parachain or token. For us, it's usually sufficient if we are able to test it on a local chain. We need to be as objective as possible when it comes to certain parachains.

**Comment by pcorrado-np** [comment]
Hi @Noc2! We updated our application:
- Fixed image
- Updated technical details section to better describe the current process
- Added a new "Our solution" section to describe our generic software solution
- Added a new "Wireframes" section to display our UI initial wireframes

Thanks again for your useful feedback!

**Review by keeganquigley — DISMISSED** [review]
Hi @pcorrado-np thanks for the changes. Looks good and it's cool to see you partnering with Oak after your previous grant. Excited to more on-chain automation tools like this. I'm happy to go ahead with it. I will mark it as ready for review and ping the rest of the committee.

**Review by Noc2 — APPROVED** [review]
Thanks for the very quick reply here and the update. I'm also happy to go ahead with it.

**Comment by pcorrado-np** [comment]
Hi @Noc2 @keeganquigley! 
Thank you both for your quick feedback. We’ll stay tuned for any news.

**Review by dsm-w3f — CHANGES_REQUESTED** [review]
@pcorrado-np thank you for the application. I think the scope of the grant interesting, however the price charged is very high. The wireframes presented has 5 pages and the price for develop them (M1) is almost 42k. Furthermore, M2 scope is to develop a smart contract, which probably you already have part of it developed in the previous grant or because of the kind of business NeoPower runs. Adding XCM to it should not be a big deal, and the price to do it is 28k. Could you consider going for a level 2 grant?

**Comment by pcorrado-np** [comment]
> @pcorrado-np thank you for the application. I think the scope of the grant interesting, however the price charged is very high. The wireframes presented has 5 pages and the price for develop them (M1) is almost 42k. Furthermore, M2 scope is to develop a smart contract, which probably you already have part of it developed in the previous grant or because of the kind of business NeoPower runs. Adding XCM to it should not be a big deal, and the price to do it is 28k. Could you consider going for a level 2 grant?

Hello @dsm-w3f, thank you for taking the time to review our application.

We truly appreciate your feedback and understand your concern regarding the pricing. 

Regarding Milestone 1, the main value of the proposal is about the Typescript logic, React hooks, and code encapsulation to ensure seamless integration with XCM in a generic way. Additionally, our proposal includes the development of the pages to create proxy accounts in a straight-forward way from both the developer and user perspective. What takes most of the estimation and effort in this milestone is the development of fully-typed reusable code for other teams.

Regarding Milestone 2, we don’t have an existing smart contract that does this. Our previous grant last year was about token streaming, and we don’t have any other smart contracts developed for Roloi so far. This development would be completely new to enable on-chain recurring transactions and integrate Kusama chains. This is important for our project because it requires robust testing since the canary DApp will interact with actual money.

From our experience while learning and teaching the Polkadot tech stack, we made sure that the estimations are accurate considering documentation, design, development, peer reviews, and testing.

We firmly believe that the tools developed through this grant will not only be valuable to our own product, to bring on-chain payment automation, but will also be beneficial to numerous other teams in the ecosystem. As part of our commitment to the community, we will actively share this in the various developers' communities we are engaged with, specifically in LATAM.

We appreciate the opportunity to apply for this grant, and we remain open to further discussions or clarifications if needed.

Thank you for your consideration, and we look forward to your response.

**Review by dsm-w3f — COMMENTED** [review]
@pcorrado-np thank you for the answer. Although your argumentation is valid, I still think it does not justify the price charged. For this reason, I won't approve this application. The other members of the committee can have a different opinion on that.

File: applications/roloi-xcm-payment-automation.md
Comment (semuelle) [inline applications/roloi-xcm-payment-automation.md:40]: Can you provide another image that would show how the workflow would be with smart contract? Who would be paying the fees for the recurring transactions, and how? Would the smart contract manage the fees per user?

**Review by semuelle — CHANGES_REQUESTED** [review]
Since I haven't used OAK before, I'm not aware how recurring transactions are created and how much of this workflow can be automated away. It seems to me that the creation manual top-up would still be necessary. If I'm reading this correct, you want to build a smart contract to simplify managing recurring payments and a UI to simplify the setup. Is that correct?

File: applications/roloi-xcm-payment-automation.md
Comment (pcorrado-np) [reply applications/roloi-xcm-payment-automation.md:40]: The workflow diagram generically shows the communication between an origin chain and OAK to enable recurring tasks, and the extrinsic can be any type of task. In our case, will be a smart contract transaction. Internally, the contract will validate the execution and trigger the configured transfer.

As for the fees, the idea is for users to top up the proxy accounts based on the defined recurrence settings. 
On the UI side, we can simulate and estimate the fees of XCM and smart contract transactions to cover an approximate number of recurring repetitions. Then we’ll suggest to the user a concrete number of tokens to transfer to the proxy account to cover fees. 
We’ll also use these estimation calculations on the UI to alert the user when they are running out of funds to pay the fees, so they can recharge the proxy account.

**Comment by pcorrado-np** [comment]
> Since I haven't used OAK before, I'm not aware how recurring transactions are created and how much of this workflow can be automated away. It seems to me that the creation manual top-up would still be necessary. If I'm reading this correct, you want to build a smart contract to simplify managing recurring payments and a UI to simplify the setup. Is that correct?

Hi, @semuelle. Thank you so much for the review!

OAK recurrently triggers XCM transactions to the origin chain according to the previously received recurring settings. At the origin chain side, a proxy account takes the transaction that includes the encoded extrinsic to be executed (in this case, the extrinsic will be a smart contract transaction).

The creation and funding of proxy accounts will be integrated into the configuration of the recurring settings and will be simplified for the user with the UI. 

And you are right: 
- The Smart Contract will be in charge of storing and validating the users recurring settings and the execution of transfers.
- The main goal of the UI is to simplify the complexities of XCM and the OAK automation process, providing a great UI/UX for configuring recurring payments.   

Thanks again for your time! We’ll stay tuned.

**Review by semuelle — COMMENTED** [review]
Thanks for the replies, @pcorrado-np. A couple more questions:

- Could you integrate the wireframes directly into the application document? Otherwise they will soon be lost.
- The wireframes show a list of "incoming" payments. Incoming where? How does that work? I assumed the list would be showing the payments set up in the smart contract.
- Using a smart contract to communicate with a pallet seems a bit... redundant. Wouldn't this architecture be possible by directly interacting with the pallet?

**Comment by pcorrado-np** [comment]
> Thanks for the replies, @pcorrado-np. A couple more questions:
> 
> * Could you integrate the wireframes directly into the application document? Otherwise they will soon be lost.
> * The wireframes show a list of "incoming" payments. Incoming where? How does that work? I assumed the list would be showing the payments set up in the smart contract.
> * Using a smart contract to communicate with a pallet seems a bit... redundant. Wouldn't this architecture be possible by directly interacting with the pallet?

Hi, @semuelle. Thanks again for your questions. We really appreciate the feedback to improve our application. 

1. Thanks! We now updated the application document by adding the wireframe images.

2. Users will be able to connect their wallets to the UI. The incoming and outgoing payments refer to the incoming and outgoing transfers of the connected wallet. Not the entire contract history. This will be useful to have a usage example of a message that selects the payment flows of a given wallet.

3. We want to leverage ink!’s capabilities to enable on-chain validation rules for any business model other teams may have. This allows for more complex models other than just scheduling token transfers. For now, the smart contract will validate that the scheduled time matches with the executed time, and not just trust on a third party message.
In the future, we also want to include the XCM communication with OAK through the contracts.

**Review by semuelle — DISMISSED** [review]
Thanks for the quick reply, @pcorrado-np. I kind of agree with Diogo that it's a fairly high price. However, I would like for this project to become the basis for many other use cases and I trust that you make it as modular and generic as you say.  

Perhaps you could add a tutorial for a second use case, e.g. how to modify the smart contract to allow token transfers.

**Review by keeganquigley — APPROVED** [review]
Thanks for the changes @pcorrado-np LGTM.

**Review by Noc2 — DISMISSED** [review]

**Review by takahser — CHANGES_REQUESTED** [review]
@pcorrado-np this looks really interesting. The proposal is well-written and describes most concepts concise and clearly. However, I do the following 2 questions:

- what's the transport mechanism you're going to use to send XCM messages, I'd assume it's HRMP? Is it roughly the same process as described in this [blog article](https://medium.com/oak-blockchain/moonbeam-oak-network-integrate-to-create-xcm-automation-1ec2b6ab5a23) that describes the integration of oak with moonriver?
- there are [various types of hooks](https://react.dev/reference/react) in react, hence the term "hook" is quite generic. Could you elaborate on the hooks you're planning to implement? Maybe some rough pseudocode could help as well, just to understand better how they're going to work.

**Comment by pcorrado-np** [comment]
> @pcorrado-np this looks really interesting. The proposal is well-written and describes most concepts concise and clearly. However, I do the following 2 questions:
> 
> * what's the transport mechanism you're going to use to send XCM messages, I'd assume it's HRMP? Is it roughly the same process as described in this [blog article](https://medium.com/oak-blockchain/moonbeam-oak-network-integrate-to-create-xcm-automation-1ec2b6ab5a23) that describes the integration of oak with moonriver?
> * there are [various types of hooks](https://react.dev/reference/react) in react, hence the term "hook" is quite generic. Could you elaborate on the hooks you're planning to implement? Maybe some rough pseudocode could help as well, just to understand better how they're going to work.
> 
> Finally, I agree that the price is rather high. But before giving a decision on my approval, I'd like to see the above unclarities resolved.

Hi @takahser. Thanks for the review!

Exactly! The process is similar and uses HRMP.
Unlike the Moonriver process, we set up the task execution and send it through XCM via UI code using Polkadot.js. When the task is triggered by OAK, a proxy account on the Astar side will execute the configured transaction. 

Regarding the hooks, we'll create our owns. Some tasks that can be encapsulated are: 
- Creating proxy accounts on both sides
- Depositing funds into these accounts to cover fees, 
- Wrapping the recurring transaction to be executed via Astar Proxy Account
- Transmitting the recurring task configuration via XCM. 

These are some of the hooks that we have in mind, but we are certain that many more will come up during development. We’ll also make sure to validate the naming and data modeling according to the standard guidelines.

For example:
carbon (2)

Thanks again!

**Review by takahser — CHANGES_REQUESTED** [review]
@pcorrado-np thanks for elaborating.

> Unlike the Moonriver process, we set up the task execution and send it through XCM via UI code using Polkadot.js. 

So the key difference between your and Moonriver's implementation would be the web UI, is that correct?

Regarding HRMP and your explanations on the React Hooks: I believe it'd be good if you could integrate this into the document. It's better to have it there since in the convo it runs at risk of getting lost.

Apart from that, I agree with @dsm-w3f that the price is rather high. I think 6 months for M1 is quite a long time for writing some hooks, the payment page which is basically a web form and some wallet integration. Especially if I consider that you're planning to have [4 developers](https://github.com/w3f/Grants-Program/pull/1857/files#diff-4fcc4da7f887d174c143505f4366c6fdb4e8f22ba02a65079ae44fbaf2e9a29aR151) working on it simultaneously while there seem to be only [2 devs](https://github.com/w3f/Grants-Program/pull/1857/files#diff-4fcc4da7f887d174c143505f4366c6fdb4e8f22ba02a65079ae44fbaf2e9a29aR117) in your team. I have a similar feeling about M2. Especially since it seems you can reuse a significant portion of the code from the Moonriver implementation. Don't get me wrong, there's nothing wrong with that, but it lowers the value your project adds to the community in contrast to a project that comes up with something completely novel.

For these reasons I'm not going to approve it either at the current price point. I think lowering it to a level 2 grant would make sense here. Given you already got 3 approvals, a level 2 version of your proposal would be approved.

**Comment by pcorrado-np** [comment]
Hi, @takahser! We greatly appreciate your review and questions. 
Here are some additional comments to offer more insights about our proposal:

> Regarding HRMP and your explanations on the React Hooks: I believe it'd be good if you could integrate this into the document. It's better to have it there since in the convo it runs at risk of getting lost.

We can update our proposal with that information. Thanks for the feedback!

> So the key difference between your and Moonriver's implementation would be the web UI, is that correct?

No, not exactly. Moonriver's implementation takes a pallet-side XCM management approach. This way a Solidity contract can create an XCM message in order to send to OAK and schedule a future transaction. OAK receives and triggers the Solidity contract message on Moonriver's side when it's time to execute the transaction.

This implementation is only useful for Solidity contract developers that will fully depend on the usage of a pallet.

With our implementation, we will provide XCM Typescript tooling for UI developers. Currently, using XCM on the UI requires a complex setup process using JavaScript.
This is actual feedback from developers that attempted to build products using XCM (including ourselves), that is why we want to improve this experience.

> Apart from that, I agree with @dsm-w3f that the price is rather high. I think 6 months for M1 is quite a long time for writing some hooks, the payment page which is basically a web form and some wallet integration. Especially if I consider that you're planning to have [4 developers](https://github.com/w3f/Grants-Program/pull/1857/files#diff-4fcc4da7f887d174c143505f4366c6fdb4e8f22ba02a65079ae44fbaf2e9a29aR151) working on it simultaneously while there seem to be only [2 devs](https://github.com/w3f/Grants-Program/pull/1857/files#diff-4fcc4da7f887d174c143505f4366c6fdb4e8f22ba02a65079ae44fbaf2e9a29aR117) in your team.

We see now that “React hooks” doesn’t describe our M1 proposal well enough. We will implement React Typescript tooling to work with XCM. This includes **hooks, types, state management and generic components**.

The UI pages are intended to showcase a use-case example of recurring payments, while our main focus is on the **reusable modules** of our project.

About our team, we are only describing our core team on the proposal. But for the FTE we are also considering designers, engineers and testers that work with us.

The estimation is **6 weeks** (not months), considering **design, development, documentation, peer reviews and testing**. Based on our experience working with cross-discipline teams and the Polkadot infrastructure, we believe that the estimation is accurate.

> I have a similar feeling about M2. Especially since it seems you can reuse a significant portion of the code from the Moonriver implementation. Don't get me wrong, there's nothing wrong with that, but it lowers the value your project adds to the community in contrast to a project that comes up with something completely novel.

Moonriver’s announcement was about opening an HRMP channel to be able to generate an XCM message and pass it to OAK at the pallet level.

That is not providing abstractions for transaction scheduling, **just a communication channel between Moonriver and OAK**.

That is why **we can’t reuse the code for M2**. We will develop an ink! smart contract that will store the scheduling configuration of each transaction, validate and execute when OAK triggers the smart contract message. This is something **completely new for the ink! ecosystem**.

> For these reasons I'm not going to approve it either at the current price point. I think lowering it to a level 2 grant would make sense here. Given you already got 3 approvals, a level 2 version of your proposal would be approved.

We understand the concern about the price point and appreciate your feedback. We also believe that some relevant technical details of our proposal weren’t clear enough based on some of the questions we received.

We genuinely believe that **this project will introduce a simpler way for developers to build automation products using Typescript + ink! + XCM**. Our main goal is to add value to the community, but we also need to make sure that our team members are fairly compensated for their work.

We can refine the description of our proposal and deliverables to provide a clearer explanation of the solution and the work required. 

**Would you reconsider your decision based on the information provided?** A level 2 grant seems too low for the work involved.

Thanks again for taking the time and effort to provide valuable feedback! 🙏🏻

We are open to providing more details and further reviewing our proposal to start working on this project.

**Comment by takahser** [comment]
@pcorrado-np

> About our team, we are only describing our core team on the proposal. But for the FTE we are also considering designers, engineers and testers that work with us.

Keep in mind that our program is tailored to support software development efforts, but not design efforts. Also, if there are more software engineers on your team, you're supposed to list them as well. Quote from our [README](https://github.com/w3f/Grants-Program/tree/82b99906c91eb627ccf8a32e49c6a2e920e4f286#wave-introduction):

> You will be asked to provide the GitHub profiles of your team members as part of your application, which we will examine for past activity and code quality.

**Review by takahser — CHANGES_REQUESTED** [review]
@pcorrado-np thanks for the follow-up comments.

> About our team, we are only describing our core team on the proposal. But for the FTE we are also considering designers, engineers and testers that work with us.

Keep in mind that our program is tailored to support software development efforts, but not design efforts. Also, if there are more software engineers on your team, you're supposed to list them as well. Quote from our [README](https://github.com/w3f/Grants-Program/tree/82b99906c91eb627ccf8a32e49c6a2e920e4f286#wave-introduction):

> You will be asked to provide the GitHub profiles of your team members as part of your application, which we will examine for past activity and code quality.

Apart from that, I still believe it's too expensive.

**Comment by pcorrado-np** [comment]
I have read and hereby sign the Contributor License Agreement.

**Comment by pcorrado-np** [comment]
Hi @dsm-w3f @takahser @Noc2 @semuelle @keeganquigley,

Thank you all for your valuable feedback.

We've decided to lower the proposal's price by removing 1 FTE for each milestone.
NeoPower will cover the costs of the remaining resources.

We are looking forward to starting working on this project.

Thanks!

**Review by takahser — CHANGES_REQUESTED** [review]
@pcorrado-np thanks for the update and lowering the price. However, I still believe a level 2 grant would be more appropriate.

Also, could you expand on the relationship between this proposal and the one you [submitted](https://github.com/OAK-Foundation/Grants-Program/pull/5) to the OAK grants program?

**Review by semuelle — APPROVED** [review]

**Review by laboon — APPROVED** [review]

**Review by SBalaguer — APPROVED** [review]

**Review by keeganquigley — APPROVED** [review]

**Comment by pcorrado-np** [comment]
To keep you informed about the status:

We are making significant progress with the Milestone 1.
We have detected a deviation from our original estimation and we are currently targeting to deliver Milestone 1 during the 2nd week of November.‌

We’ll keep track of the status through this PR.

Thank you for your support!

**Comment by keeganquigley** [comment]
Hi @pcorrado-np how is milestone 2 coming alone?

**Comment by pcorrado-np** [comment]
> Hi @pcorrado-np how is milestone 2 coming alone?

Hi @keeganquigley,

It's taking longer than expected but we are making progress with the features. Sorry for the delay and not reporting that.

We hope to reach the second milestone the week of February 5.

We'll keep you touched.

**Comment by pcorrado-np** [comment]
Hi @keeganquigley!

Status update: 
We are close to finalizing the delivery of Milestone #2 in the upcoming days. 
Everything is working as expected, and now wrapping up docs and article.

Thanks for your patience!

**Comment by keeganquigley** [comment]
Sounds good @pcorrado-np thanks for the update, much appreciated!

**Pull request merged**
